src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const stateVersions = [];
+
+  for (const action of actions) {
+    const { type, keysToRemove, extraData } = action;
+
+    switch (type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = {
+          ...newState,
+          ...extraData,
+        };
+        break;
+
+      case 'removeProperties':
+        for (const keyToRemove of keysToRemove) {
+          delete newState[keyToRemove];
+        }
```

Comment:
Consider of creating separate function for this to make `switch-case` smaller and more readable